File: packages/python/catalogmx/catalogs/sat/cfdi_4/regimen_fiscal.py
"""Catálogo c_RegimenFiscal"""

from catalogmx.catalogs.sat.cfdi_4._resolver_views import regimen_fiscal_rows


class RegimenFiscalCatalog:
    """Catálogo de Regímenes Fiscales del SAT (c_RegimenFiscal)"""

    _data: list[dict] | None = None
    _by_code: dict[str, dict] | None = None

    @classmethod
    def _load_data(cls) -> None:
        """Carga lazy desde el artefacto canónico CFDI 4.0."""
        if cls._data is None:
            cls._data = regimen_fiscal_rows()
            cls._by_code = {item["code"]: item for item in cls._data}

    @classmethod
    def get_regimen(cls, code: str) -> dict | None:
        """Obtiene un régimen fiscal por su código"""
        cls._load_data()
        return cls._by_code.get(code)

    @classmethod
    def is_valid(cls, code: str) -> bool:
        """Valida si un código de régimen fiscal es válido"""
        return cls.get_regimen(code) is not None

    @classmethod
    def is_valid_for_persona_fisica(cls, code: str) -> bool:
        """Valida si un régimen es válido para persona física"""
        regimen = cls.get_regimen(code)
        return regimen.get("fisica", False) if regimen else False

    @classmethod
    def is_valid_for_persona_moral(cls, code: str) -> bool:
        """Valida si un régimen es válido para persona moral"""
        regimen = cls.get_regimen(code)
        return regimen.get("moral", False) if regimen else False

    @classmethod
    def get_all(cls) -> list[dict]:
        """Obtiene todos los regímenes fiscales"""
        cls._load_data()
        return cls._data.copy()

File: packages/dart/lib/src/catalogs/sat/cfdi/regimen_fiscal.dart
/// Catálogo de Regímenes Fiscales (SAT)
///
/// Catálogo c_RegimenFiscal del SAT para CFDI 4.0
library;

import 'package:catalogmx/src/catalogs/base_catalog.dart';

/// Régimen Fiscal Catalog
class RegimenFiscalCatalog {
  static List<Map<String, dynamic>>? _data;
  static Map<String, Map<String, dynamic>>? _byClave;

  static void _loadData() {
    if (_data != null) return;

    final jsonData =
        BaseCatalog.loadJsonDataSync('sat/cfdi_4.0/regimen_fiscal.json');

    // Handle both list and dict formats
    if (jsonData.isNotEmpty && jsonData.first.containsKey('regimenes')) {
      _data = (jsonData.first['regimenes'] as List)
          .map((e) => e as Map<String, dynamic>)
          .toList();
    } else {
      _data = jsonData;
    }

    // Build index
    _byClave = {
      for (var item in _data!) item['clave'] as String: item,
    };
  }

  /// Obtiene todos los regímenes fiscales
  static List<Map<String, dynamic>> getAll() {
    _loadData();
    return List.from(_data!);
  }

  /// Obtiene régimen fiscal por clave
  static Map<String, dynamic>? getByClave(String clave) {
    _loadData();
    return _byClave![clave];
  }

  /// Valida si una clave de régimen fiscal existe
  static bool isValid(String clave) {
    return getByClave(clave) != null;
  }

  /// Busca regímenes fiscales por descripción
  static List<Map<String, dynamic>> search(String query) {
    _loadData();
    final queryNorm = query.toUpperCase();
    return _data!
        .where((item) =>
            (item['descripcion'] as String).toUpperCase().contains(queryNorm))
        .toList();
  }

  /// Obtiene regímenes fiscales aplicables para personas físicas
  static List<Map<String, dynamic>> getParaPersonasFisicas() {
    _loadData();
    return _data!
        .where((item) =>
            item['persona_fisica'] == true ||
            item['aplicaPersonaFisica'] == 'Sí')
        .toList();
  }

  /// Obtiene regímenes fiscales aplicables para personas morales
  static List<Map<String, dynamic>> getParaPersonasMorales() {
    _loadData();
    return _data!
        .where((item) =>
            item['persona_moral'] == true || item['aplicaPersonaMoral'] == 'Sí')
        .toList();
  }
}
